### PR TITLE
More `Num` methods

### PR DESCRIPTION
**Pull request by MarcoLizza**
Here thery are some additions I had stashed some time ago. Guess now it's time for them to be commited.

### CODE REVIEW

File: src/wren_core.c [hunk 1]
```diff
@@ -878,6 +878,24 @@ DEF_PRIMITIVE(num_cos)
   RETURN_NUM(cos(AS_NUM(args[0])));
 }
 
+DEF_PRIMITIVE(num_decimal)
```
Comment (munificent) [inline src/wren_core.c]: This isn't the right name, since it doesn't deal with decimal representation. How about `fraction`?

File: src/wren_core.c [hunk 2]
```diff
@@ -878,6 +878,24 @@ DEF_PRIMITIVE(num_cos)
...
+{
+  double value = AS_NUM(args[0]);
+  RETURN_NUM(value - (int32_t)value);
```
Comment (munificent) [inline src/wren_core.c]: This will do the wrong thing on double values that don't fit in an `int32_t`. I'd use `modf()` from math.h.

File: test/number/decimal.wren [hunk 3]
```diff
@@ -0,0 +1,8 @@
+IO.print(123.decimal)      // expect: 0
+IO.print((-123).decimal)   // expect: 0
+IO.print(0.decimal)        // expect: 0
+IO.print((-0).decimal)     // expect: -0
+IO.print(0.123.decimal)    // expect: 0.123
+IO.print(12.3.decimal)     // expect: 0.3
+IO.print((-0.123).decimal) // expect: -0.123
+IO.print((-12.3).decimal)  // expect: -0.3
```
Comment (munificent) [inline test/number/decimal.wren]: Test a really big number that doesn't fit in an int.

File: src/wren_core.c [hunk 4]
```diff
@@ -878,6 +878,24 @@ DEF_PRIMITIVE(num_cos)
...
+}
+
+DEF_PRIMITIVE(num_deg)
```
Comment (munificent) [inline src/wren_core.c]: Maybe it's just me, but I don't think this is that useful. Leave it out?

File: src/wren_core.c [hunk 5]
```diff
@@ -878,6 +878,24 @@ DEF_PRIMITIVE(num_cos)
...
+{
+  RETURN_NUM(floor(AS_NUM(args[0]) * 57.2957795130823208768));
+}
+
+DEF_PRIMITIVE(num_div)
+{
+  int32_t left = (int32_t)AS_NUM(args[0]);
+  int32_t right = (int32_t)AS_NUM(args[1]);
+  RETURN_NUM(left / right);
```
Comment (munificent) [inline src/wren_core.c]: Like `decimal`, this will also do the wrong thing on > 32-bit values.

That may be OK, if this is what the user wants. I guess it's a question of is this for forced-32-bit integer divide, or just "how many times does the RHS completely divide the LHS?" My intuition is the latter in which case I think you want to just divide the args as doubles and then use `fmod()` to truncate the result.

File: src/wren_core.c [hunk 6]
```diff
@@ -888,6 +906,26 @@ DEF_PRIMITIVE(num_isNan)
   RETURN_BOOL(isnan(AS_NUM(args[0])));
 }
 
+DEF_PRIMITIVE(num_rad)
```
Comment (munificent) [inline src/wren_core.c]: Ditto for `deg`. I don't feel this is that useful.

File: src/wren_core.c [hunk 7]
```diff
@@ -888,6 +906,26 @@ DEF_PRIMITIVE(num_isNan)
...
+{
+  RETURN_NUM(floor(AS_NUM(args[0]) / 57.2957795130823208768));
+}
+
+DEF_PRIMITIVE(num_sign)
+{
+  double value = AS_NUM(args[0]);
+  double sign;
+  if (value > 0) {
```
Comment (munificent) [inline src/wren_core.c]: Style nits: Can you put the `{` on the next line?

File: src/wren_core.c [hunk 8]
```diff
@@ -888,6 +906,26 @@ DEF_PRIMITIVE(num_isNan)
...
+    sign = 1;
+  } else
+  if (value < 0) {
```
Comment (munificent) [inline src/wren_core.c]: ...and put this on the same line as the `else`?

File: src/wren_core.c [hunk 9]
```diff
@@ -888,6 +906,26 @@ DEF_PRIMITIVE(num_isNan)
...
 
+DEF_PRIMITIVE(num_rad)
+{
+  RETURN_NUM(floor(AS_NUM(args[0]) / 57.2957795130823208768));
+}
+
+DEF_PRIMITIVE(num_sign)
+{
+  double value = AS_NUM(args[0]);
+  double sign;
```
Comment (munificent) [inline src/wren_core.c]: Instead of using a temp variable, I'd probably just return directly from within each branch.

File: src/wren_core.c [hunk 10]
```diff
@@ -928,6 +966,11 @@ DEF_PRIMITIVE(num_toString)
   RETURN_VAL(wrenNewString(vm, buffer, length));
 }
 
+DEF_PRIMITIVE(num_truncate)
+{
+  RETURN_NUM((int32_t)(AS_NUM(args[0])));
```
Comment (munificent) [inline src/wren_core.c]: `fmod()` is your friend here too. :)

File: test/number/div.wren [hunk 11]
```diff
@@ -0,0 +1,5 @@
+IO.print((8).div(2))   // expect: 4
+IO.print((8.5).div(2)) // expect: 4
+IO.print((9).div(2))   // expect: 4
+IO.print((9.9).div(2)) // expect: 4
+IO.print((10).div(2))  // expect: 5
```
Comment (munificent) [inline test/number/div.wren]: More stuff to test:

* Large number
* Zero LHS.
* Zero RHS.
* Floating point RHS.
* Negative arguments.

File: test/number/sign.wren [hunk 12]
```diff
@@ -0,0 +1,6 @@
+IO.print(123.sign)      // expect: 1
+IO.print((-123).sign)   // expect: -1
+IO.print(0.sign)        // expect: 0
+IO.print((-0).sign)     // expect: 0
```
Comment (munificent) [inline test/number/sign.wren]: Good corner case test. Is this the result as user would expect, though?

File: test/number/truncate.wren [hunk 13]
```diff
@@ -0,0 +1,8 @@
+IO.print(123.truncate)      // expect: 123
+IO.print((-123).truncate)   // expect: -123
+IO.print(0.truncate)        // expect: 0
+IO.print((-0).truncate)     // expect: 0
+IO.print(0.123.truncate)    // expect: 0
+IO.print(12.3.truncate)     // expect: 12
+IO.print((-0.123).truncate) // expect: 0
+IO.print((-12.3).truncate)  // expect: -12
```
Comment (munificent) [inline test/number/truncate.wren]: Test a large number here too.

**Comment by munificent** [comment]
I added lots of comments, but thank you for doing these!

File: src/wren_core.c (hunk 2 above)
Comment (MarcoLizza) [inline src/wren_core.c]: Yep. I agree, in fact was my first option...

File: src/wren_core.c (hunk 1 above)
Comment (MarcoLizza) [inline src/wren_core.c]: Yup. Agreed.

File: src/wren_core.c (hunk 4 above)
Comment (MarcoLizza) [inline src/wren_core.c]: In fact I wasn't really sure both `deg` and `rad` are so useful for everybody (due to the fact that they could be easily computed in Wren itself), but due to the presence of the trigonometric methods, I thought it would be funny to write something like
```
270.rad.sin
```
:)

File: src/wren_core.c (hunk 6 above)
Comment (MarcoLizza) [inline src/wren_core.c]: Fine.

File: src/wren_core.c (hunk 7 above)
Comment (MarcoLizza) [inline src/wren_core.c]: Yep, sorry.

I tend to adhere to the Wren coding-styles but from time to time other projects (and/or personal) style crawls into. :)

File: src/wren_core.c (hunk 9 above)
Comment (MarcoLizza) [inline src/wren_core.c]: I tend not to use multiple returns but will fix this. :+1:

File: test/number/sign.wren (hunk 12 above)
Comment (MarcoLizza) [inline test/number/sign.wren]: According to the definition of sign-function, yes that's what I expect to return. In fact, `0` and `-0` are both, ehm, zero... hence their sign is zero.

What's your opinion?

**Comment by MarcoLizza** [comment]
Got a lot of homework to do. Better roll up my sleeves and do them. :+1:

File: src/wren_core.c (hunk 5 above)
Comment (MarcoLizza) [inline src/wren_core.c]: In my plans this was meant to be a integer division, that is the same as
```
lsh.truncate / rhs.truncate
```
and not
```
(lhs / rhs).truncate
```
However, it now occurs to me that since both the division types can be easily obtained in Wren itself (now that we have the `truncate` method) perhaps the `div` method is useless.

Let's keep it simple, as usual, and leave it out. We would be forcing an (arbitrary) integer-division method when one may prefer the other and compute it by himself in Wren.

What do you think?

File: test/number/sign.wren (hunk 12 above)
Comment (munificent) [inline test/number/sign.wren]: Yeah, I think I'd expect it to follow `signbit()`. Negative zero does feel pretty squishy to me, but IEEE doubles support it and Wren follows that so my normal guideline is to just assume the C standard lib folks have put more thought into this than I have so just follow what they do. :)

Comment (MarcoLizza) [inline test/number/sign.wren]: Agreed, I'll fix this evening...

**Comment by munificent** [comment]
:D

**Comment by MarcoLizza** [comment]
Still missing the sign change to adehere with signbit. Will throw a pull request to change this. :+1:

**Comment by munificent** [comment]
Sounds good. I'd say we should just *use* `signbit()`, but I think it's C99-only so I'm hesitant to depend on that. :confused:

**Comment by MarcoLizza** [comment]
Yep, in fact `signbit` is C99/C++11 only.

Perhaps we could find our way by mean of a union/cast/masking? Sounds too hacky?

**Comment by munificent** [comment]
Yeah, I'm happy with the solution you already did. :)

**Comment by MarcoLizza** [comment]
All rights, let's keep this then! =D
